File: partition-list.py
class Solution(object):
    def partition(self, head, x):
        # create two linked list: lz and ge
        lzTail, geTail = lzHead, geHead = ListNode(0), ListNode(0)

        # split input into these two list
        while head:
            if head.val < x:
                lzTail.next, lzTail = head, head
            else:
                geTail.next, geTail = head, head
            head = head.next

        # concatenate the two list, make sure None is at the end of the list
        lzTail.next, geTail.next = geHead.next, None

        return lzHead.next

File: partition-list.java
import java.util.Arrays;

public class Solution {
  // public class ListNode {
  //   int val;
  //   ListNode next;
  //   ListNode(int x) { val = x; }
  // }

  public ListNode partition(ListNode head, int x) {
    ListNode hat = new ListNode(0);
    hat.next = head;
    ListNode lessThanEnd = hat;
    ListNode largerThanEnd = hat;
    ListNode cur = head;
    while (cur != null) {
      if (cur.val >= x) {
        largerThanEnd = cur;
        cur = cur.next;
      } else if (lessThanEnd.next == cur) {
        lessThanEnd = cur;
        cur = cur.next;
      } else {
        ListNode nextLessThanEnd = cur;
        cur = cur.next;
        nextLessThanEnd.next = lessThanEnd.next;
        lessThanEnd.next = nextLessThanEnd;
        lessThanEnd = nextLessThanEnd;
        largerThanEnd.next = cur;
      }
    }

    return hat.next;
  }

  public static void main (String[] args) {
    Solution s = new Solution();
  }
}
